0.14.0 - Doesn't work without For

Hi.

I tried to use new behavior when we don't need to specify `For` and it looks like it doesn't work.
It is enough to have only one controller in a project without `For` to break it.

What I did:
```
kubebuilder init --domain my.domain --repo my.domain/guestbook --plugins=go/v4-alpha
kubebuilder create api --group webapp --version v1 --kind Guestbook # accept creating api/controller
go get sigs.k8s.io/controller-runtime
go mod tidy
```

Changed `SetupWithManager` a bit:
```
// SetupWithManager sets up the controller with the Manager.
func (r *GuestbookReconciler) SetupWithManager(mgr ctrl.Manager) error {
	q := make(chan event.GenericEvent)
	return ctrl.NewControllerManagedBy(mgr).
		Watches(&source.Channel{Source: q}, &handler.EnqueueRequestForObject{}).
		Named("test").
		Complete(r)
}
```

When I run it, I receive the following error:
```
Got a connection, launched process /private/var/folders/_r/8rbt36_d7fj5m304m1xfm31r0000gn/T/GoLand/___go_build_my_domain_guestbook (pid = 72376).
2022-12-15T09:30:08+03:00       INFO    controller-runtime.metrics      Metrics server is starting to listen    {"addr": ":8080"}
2022-12-15T09:30:08+03:00       INFO    setup   starting manager
2022-12-15T09:30:08+03:00       INFO    Starting server {"path": "/metrics", "kind": "metrics", "addr": "[::]:8080"}
2022-12-15T09:30:08+03:00       INFO    Starting server {"kind": "health probe", "addr": "[::]:8081"}
2022-12-15T09:30:08+03:00       INFO    Starting EventSource    {"controller": "test", "source": "kind source: unknown type"}
2022-12-15T09:30:08+03:00       INFO    Stopping and waiting for non leader election runnables
2022-12-15T09:30:08+03:00       INFO    Stopping and waiting for leader election runnables
2022-12-15T09:30:08+03:00       INFO    Stopping and waiting for caches
2022-12-15T09:30:08+03:00       INFO    Stopping and waiting for webhooks
2022-12-15T09:30:08+03:00       INFO    Wait completed, proceeding to shutdown the manager
2022-12-15T09:30:08+03:00       ERROR   setup   problem running manager {"error": "must specify Kind.Type"}
```


If I try to use this behaviour in more complex operator, where I have more than one controller, I receive the following output (panic comes in 30sec after start)
```
2022-12-15T09:31:10.888+0300    INFO    controller-runtime.metrics      metrics/listener.go:44  Metrics server is starting to listen    {"addr": ":8080"}
2022-12-15T09:31:10.889+0300    INFO    setup   sbc-operator/main.go:113        starting manager
2022-12-15T09:31:10.889+0300    INFO    manager/internal.go:369 Starting server {"path": "/metrics", "kind": "metrics", "addr": "[::]:8080"}
2022-12-15T09:31:10.889+0300    INFO    manager/internal.go:369 Starting server {"kind": "health probe", "addr": "[::]:8081"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:186    Starting EventSource    {"controller": "pod", "controllerGroup": "", "controllerKind": "Pod", "source": "kind source: *v1.Pod"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:186    Starting EventSource    {"controller": "garbage", "source": "kind source: unknown type"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:186    Starting EventSource    {"controller": "pod", "controllerGroup": "", "controllerKind": "Pod", "source": "kind source: *v1.Pod"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:194    Starting Controller     {"controller": "pod", "controllerGroup": "", "controllerKind": "Pod"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:186    Starting EventSource    {"controller": "node", "controllerGroup": "", "controllerKind": "Node", "source": "kind source: *v1.Node"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:228    Starting workers        {"controller": "pod", "controllerGroup": "", "controllerKind": "Pod", "worker count": 10}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:194    Starting Controller     {"controller": "node", "controllerGroup": "", "controllerKind": "Node"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:194    Starting Controller     {"controller": "pod", "controllerGroup": "", "controllerKind": "Pod"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:228    Starting workers        {"controller": "pod", "controllerGroup": "", "controllerKind": "Pod", "worker count": 1}
2022-12-15T09:31:10.991+0300    INFO    manager/internal.go:576 Stopping and waiting for non leader election runnables
2022-12-15T09:31:10.991+0300    INFO    manager/internal.go:580 Stopping and waiting for leader election runnables
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:248    Shutdown signal received, waiting for all workers to finish     {"controller": "pod", "controllerGroup": "", "controllerKind": "Pod"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:228    Starting workers        {"controller": "node", "controllerGroup": "", "controllerKind": "Node", "worker count": 1}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:250    All workers finished    {"controller": "pod", "controllerGroup": "", "controllerKind": "Pod"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:248    Shutdown signal received, waiting for all workers to finish     {"controller": "pod", "controllerGroup": "", "controllerKind": "Pod"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:250    All workers finished    {"controller": "pod", "controllerGroup": "", "controllerKind": "Pod"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:248    Shutdown signal received, waiting for all workers to finish     {"controller": "node", "controllerGroup": "", "controllerKind": "Node"}
2022-12-15T09:31:10.991+0300    INFO    controller/controller.go:250    All workers finished    {"controller": "node", "controllerGroup": "", "controllerKind": "Node"}
2022-12-15T09:31:10.991+0300    ERROR   controller-runtime.source       source/source.go:148    failed to get informer from cache       {"error": "Timeout: failed waiting for *v1.Node Informer to sync"}
sigs.k8s.io/controller-runtime/pkg/source.(*Kind).Start.func1.1
        /Users/alexandr/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.0/pkg/source/source.go:148
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext
        /Users/alexandr/go/pkg/mod/k8s.io/apimachinery@v0.26.0/pkg/util/wait/wait.go:235
k8s.io/apimachinery/pkg/util/wait.poll
        /Users/alexandr/go/pkg/mod/k8s.io/apimachinery@v0.26.0/pkg/util/wait/wait.go:582
k8s.io/apimachinery/pkg/util/wait.PollImmediateUntilWithContext
        /Users/alexandr/go/pkg/mod/k8s.io/apimachinery@v0.26.0/pkg/util/wait/wait.go:547
sigs.k8s.io/controller-runtime/pkg/source.(*Kind).Start.func1
        /Users/alexandr/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.0/pkg/source/source.go:136
2022-12-15T09:31:40.992+0300    INFO    manager/internal.go:586 Stopping and waiting for caches
2022-12-15T09:32:09.435+0300    INFO    manager/internal.go:590 Stopping and waiting for webhooks
panic: [must specify Kind.Type, failed waiting for all runnables to end within grace period of 30s: context deadline exceeded]
```

--- a/pkg/builder/controller.go
+++ b/pkg/builder/controller.go
@@ -216,15 +216,17 @@ func (blder *Builder) project(obj client.Object, proj objectProjection) (client.
 
 func (blder *Builder) doWatch() error {
 	// Reconcile type
-	typeForSrc, err := blder.project(blder.forInput.object, blder.forInput.objectProjection)
-	if err != nil {
-		return err
-	}
-	src := &source.Kind{Type: typeForSrc}
-	hdler := &handler.EnqueueRequestForObject{}
-	allPredicates := append(blder.globalPredicates, blder.forInput.predicates...)
-	if err := blder.ctrl.Watch(src, hdler, allPredicates...); err != nil {
-		return err
+	if blder.forInput.object != nil {
+		typeForSrc, err := blder.project(blder.forInput.object, blder.forInput.objectProjection)
+		if err != nil {
+			return err
+		}
+		src := &source.Kind{Type: typeForSrc}
+		hdler := &handler.EnqueueRequestForObject{}
+		allPredicates := append(blder.globalPredicates, blder.forInput.predicates...)
+		if err := blder.ctrl.Watch(src, hdler, allPredicates...); err != nil {
+			return err
+		}
 	}
 
 	// Watches the managed types

--- a/pkg/builder/controller_test.go
+++ b/pkg/builder/controller_test.go
@@ -376,6 +376,23 @@ var _ = Describe("application", func() {
 			defer cancel()
 			doReconcileTest(ctx, "4", m, true, bldr)
 		})
+
+		It("should Reconcile without For", func() {
+			m, err := manager.New(cfg, manager.Options{})
+			Expect(err).NotTo(HaveOccurred())
+
+			bldr := ControllerManagedBy(m).
+				Named("Deployment").
+				Watches( // Equivalent of For
+						&source.Kind{Type: &appsv1.Deployment{}}, &handler.EnqueueRequestForObject{}).
+				Watches( // Equivalent of Owns
+					&source.Kind{Type: &appsv1.ReplicaSet{}},
+					&handler.EnqueueRequestForOwner{OwnerType: &appsv1.Deployment{}, IsController: true})
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			doReconcileTest(ctx, "9", m, true, bldr)
+		})
 	})
 
 	Describe("Set custom predicates", func() {

